client/src/components/Atoms/CodeInput.tsx

```diff
@@ -0,0 +1,114 @@
+import React, { useEffect } from 'react'
+import styled from 'styled-components'
+import { Classes } from '@blueprintjs/core'
+import { replaceAtIndex, range } from '../../utils/array'
+import { Row } from './Layout'
+import { assert } from '../utilities'
+
+const DigitInput = styled.input.attrs({
+  className: Classes.INPUT,
+  type: 'text',
+})`
+  text-align: center;
+`
+
+// We only support a small subset of the HTMLInputElement interface to start,
+// but this can be expanded as needed.
+type InputProps = Pick<
+  React.InputHTMLAttributes<HTMLInputElement>,
+  'name' | 'id'
+>
+
+interface ICodeInputProps extends InputProps {
+  length: number
+  value: string
+  onChange: (value: string) => void
+}
+
+/**
+ * CodeInput takes the same props as a controlled text input, but
+ * actually renders and coordinates multiple individual digit inputs.
+ *
+ * Example:
+ *  const [value, setValue] = useState('')
+ *  <CodeInput length={3} value={value} onChange={setValue} />
+ */
+const CodeInput: React.FC<ICodeInputProps> = ({
+  length,
+  value = '',
+  onChange,
+  ...htmlInputProps
+}) => {
+  assert(/^\d*$/.test(value), 'CodeInput value must be a string of digits')
+
+  const digitInputRefs = range(0, length - 1).map(() =>
+    React.createRef<HTMLInputElement>()
+  )
+
+  const focusDigitInput = (index: number) => {
+    const digitInputRef = digitInputRefs[index]
+    if (digitInputRef?.current) {
+      digitInputRef.current.focus()
+    }
+  }
+
+  const moveFocusRight = (index: number) => {
+    if (index < length - 1) {
+      focusDigitInput(index + 1)
+    }
+  }
+
+  const moveFocusLeft = (index: number) => {
+    if (index > 0) {
+      focusDigitInput(index - 1)
+    }
+  }
+
+  const onDigitKeyDown = (index: number, key: string) => {
+    const firstEmptyDigitIndex = value.length
+    if (key.match(/[0-9]/)) {
+      if (index === firstEmptyDigitIndex) {
+        onChange(replaceAtIndex(value.split(''), index, key).join(''))
+        moveFocusRight(index)
+      }
+    } else if (key === 'Backspace') {
+      onChange(replaceAtIndex(value.split(''), index, '').join(''))
+      moveFocusLeft(index)
+    } else if (key === 'ArrowLeft') {
+      moveFocusLeft(index)
+    } else if (key === 'ArrowRight') {
+      if (index < firstEmptyDigitIndex) {
+        moveFocusRight(index)
+      }
+    }
+  }
+
+  // Whenever we have no digits entered, focus the first digit input
+  // E.g. on mount, after backspacing, or after form reset
+  useEffect(() => {
+    if (value === '') {
+      focusDigitInput(0)
+    }
+  })
```

Comment:
I believe, reading the comment, that this is intentional, but just sanity checking that we're intentionally not including a dependency array here and running this hook on every render

Reply:
Good question. I have been trying out never including a dependency array for `useEffect` hooks to make sure that I'm not relying on that performance optimization for correctness.

Do you think there's a performance reason to include one here? I could think of a few cases where there's a rerender and `value` stays `''` where we could avoid refocusing, but it seems likely negligible to me.

Reply:
> I have been trying out never including a dependency array for useEffect hooks to make sure that I'm not relying on that performance optimization for correctness

Very interesting!

Agreed that the performance concern here is negligible